src/ucp/wireup/wireup_ep.c

```diff
@@ -579,6 +579,243 @@ ssize_t ucp_wireup_ep_sockaddr_fill_private_data(void *arg, const char *dev_name
+
+    if (ucp_worker_iface_is_tl_p2p(&wireup_ep->sockaddr_tl_wiface->attr)) {
+        status = ucp_wireup_ep_connect_to_ep(&wireup_ep->super.super,
+                                             remote_data->dev_addr,
+                                             addr.address_list[0].ep_addr);
+    } else {
+        uct_ep_params_t params;
+
+        ucs_assert(wireup_ep->sockaddr_tl_wiface->attr.cap.flags &
+                   UCT_IFACE_FLAG_CONNECT_TO_IFACE);
+
+        params.field_mask = UCT_EP_PARAM_FIELD_IFACE    |
+                            UCT_EP_PARAM_FIELD_DEV_ADDR |
+                            UCT_EP_PARAM_FIELD_IFACE_ADDR;
+
+        params.iface      = wireup_ep->sockaddr_tl_wiface->iface;
+        params.dev_addr   = remote_data->dev_addr;
+        params.iface_addr = addr.address_list[0].iface_addr;
+        status = uct_ep_create(&params, &wireup_ep->super.uct_ep);
+        wireup_ep->flags |= UCP_WIREUP_EP_FLAG_LOCAL_CONNECTED;
+    }
+
+    if (status != UCS_OK) {
+        goto free_addr_list;
+    }
+
+    ucp_ep_sockaddr_cm_wireup_complete(ucp_ep, ep);
+
+free_addr_list:
+    ucs_free(addr.address_list);
+out:
+    if (status == UCS_OK) {
+        return;
+    }
+err_out:
+    ucp_worker_set_ep_failed(ucp_ep->worker, ucp_ep, &wireup_ep->super.super,
+                             wireup_idx, status);
+}
+
+ucs_status_t ucp_wireup_ep_cm_connect_to_sockaddr(uct_ep_h uct_ep,
+                                                  const ucp_ep_params_t *params)
+{
+    ucp_wireup_ep_t *wireup_ep = ucs_derived_of(uct_ep, ucp_wireup_ep_t);
+    ucp_ep_h ucp_ep            = wireup_ep->super.ucp_ep;
+    ucp_worker_h worker        = ucp_ep->worker;
+    uct_ep_params_t cm_lane_params;
+    ucs_status_t status;
+
+    cm_lane_params.field_mask = UCT_EP_PARAM_FIELD_CM                    |
+                                UCT_EP_PARAM_FIELD_USER_DATA             |
+                                UCT_EP_PARAM_FIELD_SOCKADDR              |
+                                UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS     |
+                                UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB      |
+                                UCT_EP_PARAM_FIELD_SOCKADDR_CONNECT_CB   |
+                                UCT_EP_PARAM_FIELD_SOCKADDR_DISCONNECT_CB;
+
+    cm_lane_params.user_data                  = ucp_ep;
+    cm_lane_params.sockaddr                   = &params->sockaddr;
+    cm_lane_params.sockaddr_cb_flags          = UCT_CB_FLAG_ASYNC;
+    cm_lane_params.sockaddr_pack_cb           = ucp_wireup_sockaddr_cm_priv_pack_cb;
+    cm_lane_params.sockaddr_connect_cb.client = ucp_wireup_sockaddr_client_connect_cb;
+    cm_lane_params.disconnect_cb              = ucp_ep_sockaddr_disconnect_cb;
+
+    ucs_assert_always(ucp_worker_num_cm_cmpts(worker) == 1);
+    cm_lane_params.cm = worker->cms[0].cm;
+    status = uct_ep_create(&cm_lane_params, &wireup_ep->sockaddr_ep);
+
+    ucp_ep->flags |= UCP_EP_FLAG_LOCAL_CONNECTED;
```

Comment:
what if status isn't UCS_OK?

Reply:
the err handling will be invoked below (`ucp_worker_set_ep_failed`)

Reply:
but why is the LOCAL_CONNECTED flag set if status may be not UCS_OK?

Reply:
OK, will wrap up with UCS_OK check

Reply:
done